### PR TITLE
feat: pluggable generation-backend abstraction (#88, step 1 of 2)

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary

Extracts the single `HF_model.generate` call in `Model._generate_chunk` into a `GenerationBackend` Protocol. This is **almost entirely a refactor** — one intentional, documented semantic change noted below. PR 1 of 2 toward SGLang integration (#88).

## What changed

- `src/MEDS_EIC_AR/model/backends/base.py` — `GenerationBackend` Protocol (runtime_checkable) with one method: `generate_chunk(input_ids, *, attention_mask, generation_config, **kwargs) -> Tensor`. Returns **new tokens only**; prompt-slicing is the backend's responsibility so "new-tokens-only" engines like SGLang don't need to reassemble a padded full output. Protocol docstring documents the per-row post-EOS padding invariant the rolling loop relies on.
- `src/MEDS_EIC_AR/model/backends/hf.py` — `HFBackend`, the default backend. Forwards calls to `hf_model.generate` and slices off the prompt.
- `Model.__init__` sets `self._backend = HFBackend(self.HF_model)`. `set_backend()` allows swapping for tests / future backends (e.g. SGLang in PR 2).
- `Model._generate_chunk` delegates to `self._backend.generate_chunk` — one-line swap. `GenerationConfig` construction, rolling loop, EOS mask, and prompt-slicing convention are all unchanged.
- `tests/test_generation_backend.py` — 6 unit tests covering default wiring, prompt slicing, backend redirection, structural protocol compliance, and the two kwarg-filter branches.

## One intentional semantic change: `**kwargs` filtering in `HFBackend`

This was flagged by reviewers as worth calling out explicitly. The old code in `Model._generate_chunk` forwarded `**kwargs` directly to `HF_model.generate`; the new `HFBackend.generate_chunk` filters first, and this matters:

- If HF's `generate` has a `**kwargs` VAR_KEYWORD parameter (**the case on current transformers — verified**), `HFBackend` forwards all caller kwargs unchanged and lets HF validate. Dynamic-kwarg generation controls like `temperature` / `top_p` / `top_k` that HF accepts via VAR_KEYWORD (not as explicit named params) are preserved.
- If HF's `generate` has no VAR_KEYWORD (a hypothetical future transformers layout), `HFBackend` falls back to name-filtering kwargs against the signature's explicit parameter list — so callers that have passed a kwarg only meaningful to a *different* backend (e.g. a future SGLang-specific throughput knob) get the unknown key silently stripped rather than hitting an HF `TypeError`.

Net effect on today's callers: **no behavioral change** — the VAR_KEYWORD branch is what runs, and it's a straight pass-through. The change is forward-looking: once PR 2 adds SGLang, multi-backend callers can pass a union of kwargs and let each backend take only what it knows.

Protocol requires backends to satisfy this contract (only forward engine-accepted options); `HFBackend` does it via signature inspection, a future `SGLangBackend` will do it via a hand-maintained strip list because SGLang's engine accepts a narrower set.

## Why new-tokens-only at the backend boundary

SGLang's native output is a list of new token ids per request, not a padded "prompt + new" tensor. Returning the full padded output would force every SGLang call to allocate a max-length tensor, copy the prompt back in, and pad — pure waste. The slice inside the adapter costs HF nothing (it was already slicing) and saves SGLang the reassembly.

## Test plan

- [x] `uv run pytest tests/test_generation_backend.py` — 6/6 pass (4 structural + 2 for the kwarg-filter branches)
- [x] Full grammar CLI test suite runs against the backend via the merged `tests/grammar/` — passes end-to-end
- [x] CI green on the latest commit
- [x] `uv run pre-commit run` clean

### CODE REVIEW

File: tests/test_generation_backend.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        do_sample=False,
+        num_beams=1,
+        pad_token_id=0,
+        eos_token_id=model.HF_model.config.eos_token_id,
+    )
+    out = model.backend.generate_chunk(input_ids, attention_mask=attention_mask, generation_config=cfg)
+
+    # New-tokens-only shape: batch preserved, length equals the per-call budget (no early EOS
+    # expected from a randomly initialized tiny model in 2 steps, so this is exact).
+    assert out.shape == (2, 2)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_generation_backend.py]: Fixed in 85ccac2 by comparing against a direct `HF_model.generate(...)` call sliced by the prompt length (your suggested shape), plus `shape[1] <= max_new_tokens` to stay robust to early EOS. Verified the adapter is a no-op on top of the underlying call.

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+The point of this layer is to isolate *where* a per-chunk generate call executes (HuggingFace
+today, SGLang or another engine in the future — see issue #88) from *how* the surrounding rolling
+sliding-window loop, EOS bookkeeping, and token accounting work. Only the innermost
+``generate_chunk`` call is swappable; everything else stays in :class:`MEDS_EIC_AR.model.model.Model`.
+
+A backend implementation takes a right-padded prompt tensor plus the HF-style ``GenerationConfig``
+the caller has already built and returns the newly generated tokens (the portion of the HF output
+that comes *after* the prompt). The slicing is deliberately the backend's responsibility: engines
+like SGLang produce "new tokens only" natively, and pushing the slice into the adapter keeps the
+calling code identical across backends.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Valid catch — confirmed this repo uses `padding_side: LEFT` (see `configs/datamodule/generate_trajectories.yaml:9`). Fixed in 85ccac2: the top-level docstring now says "padded token IDs" and notes both the left-pad convention and the post-EOS right-padding that rolling chunks can carry, deferring to `attention_mask` for identifying real tokens.

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        """Run one generate pass and return only the newly generated tokens.
+
+        Args:
+            input_ids: ``[B, L_in]`` right-padded prompt tokens.
+            attention_mask: Optional ``[B, L_in]`` attention mask; ``True`` for real prompt tokens.
+            generation_config: A fully-populated :class:`transformers.GenerationConfig` describing
+                per-call budget (``max_new_tokens``), sampling mode, pad/EOS ids, etc.
+            **kwargs: Forwarded to the underlying engine. Callers pass backend-specific options
+                here (e.g. HF ``logits_processor``); backends must tolerate unknown keys by
+                silently passing them through to the engine.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Good catch — fixed in 85ccac2. The contract now reads: "implementations must only forward options supported by the active engine and silently ignore or strip the rest". Matches your suggested wording, and it is internally consistent with `transformers.GenerationMixin.generate` raising on unknown kwargs.

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        *,
+        attention_mask: torch.Tensor | None,
+        generation_config: GenerationConfig,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Run one generate pass and return only the newly generated tokens.
+
+        Args:
+            input_ids: ``[B, L_in]`` right-padded prompt tokens.
+            attention_mask: Optional ``[B, L_in]`` attention mask; ``True`` for real prompt tokens.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Fixed in 85ccac2 — same change as the top-level-docstring comment. The method docstring now says "padded token IDs. Padding direction is caller-defined" and points at `attention_mask` as the authoritative signal for real-token positions.

**Comment by mmcdermott** [comment]
@copilot please review this PR

File: src/MEDS_EIC_AR/model/backends/hf.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        attention_mask: torch.Tensor | None,
+        generation_config: GenerationConfig,
+        **kwargs,
+    ) -> torch.Tensor:
+        out = self.hf_model.generate(
+            input_ids,
+            attention_mask=attention_mask,
+            generation_config=generation_config,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/hf.py:72]: Addressed via the `inspect.signature`-based filter landed in 4cdd37e + the perf cache in e454eda (signature lookup cached at `__init__` time to avoid recomputing per call). `HFBackend.generate_chunk` now filters `**kwargs` to HF `generate`-accepted parameter names before forwarding, so backend-agnostic keys from multi-backend callers get silently dropped rather than raising. Docstring in 453e132 documents the policy explicitly. Satisfies the protocol.

File: tests/test_generation_backend.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
+"""Unit tests for the generation-backend abstraction (issue #88, step 1 of 2).
+
+These tests only exercise the :class:`GenerationBackend` protocol contract and the
+:class:`HFBackend` adapter. They deliberately *do not* drive the full CLI pipeline — that's
+covered by ``test_generate_trajectories.py`` and ``test_pattern_generation_cli.py``. The goal
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_generation_backend.py]: Addressed. The file moved to `tests/grammar/test_cli.py` in the dev merge; docstring now points there (829515f). The literal `tests/test_pattern_generation.py` path in the original suggestion is now also stale — the in-process tests live at `tests/grammar/test_in_process.py`.

File: src/MEDS_EIC_AR/model/backends/hf.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        self,
+        input_ids: torch.Tensor,
+        *,
+        attention_mask: torch.Tensor | None,
+        generation_config: GenerationConfig,
+        **kwargs,
+    ) -> torch.Tensor:
+        if kwargs:
+            kwargs = {k: v for k, v in kwargs.items() if k in self._supported_generate_params}
+        out = self.hf_model.generate(
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/hf.py:67]: Good catch — you are right, and the filter as-landed was broken. Verified locally: `inspect.signature(hf_model.generate).parameters` returns 13 entries, 12 POSITIONAL_OR_KEYWORD plus one VAR_KEYWORD (`kwargs`). Options like `temperature`, `top_p`, `top_k`, `renormalize_logits` flow through the VAR_KEYWORD and are validated dynamically inside `generate` against `GenerationConfig` fields — they are NOT named parameters, so the `k in parameters` filter dropped them silently.

Fixed in 6fa671d. `HFBackend.__init__` now detects whether `generate` has a VAR_KEYWORD. If yes (current transformers: always does), it skips filtering entirely and lets HF validate; truly-unknown kwargs get the normal HF `TypeError`. Fall-back path filters against named params only if a hypothetical future transformers drops the VAR_KEYWORD. Docstring updated to explain the signature-shape branch.

Decided against also layering in `generation_config.to_dict()` keys + `hf_model.forward` params — that union overlaps with what the VAR_KEYWORD already accepts in practice, and constructing it at init time adds complexity for no behavioral win over "let HF do the validation."

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            **kwargs: Backend-specific per-call options. Callers may provide keys that are only
+                meaningful to some backends (e.g. HF ``logits_processor``); implementations must
+                only forward options supported by the active engine and silently ignore or strip
+                the rest, since e.g. ``transformers.GenerationMixin.generate`` raises on unknown
+                kwargs.
+
+        Returns:
+            A ``[B, new_len]`` tensor of newly generated tokens, with the prompt slice already
+            stripped. ``new_len`` is ``generation_config.max_new_tokens`` or less if the engine
+            emitted EOS earlier.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Good point — the invariant was implicit (and HF-specific) where it needed to be explicit. Added in 6fa671d, with your suggested language: the protocol return-value docstring now states that post-EOS positions within a chunk must be filled with `generation_config.pad_token_id` (or the row may be truncated). Also notes the rolling loop depends on this and that HFBackend satisfies it natively, but non-HF backends must honor it explicitly.

File: src/MEDS_EIC_AR/model/model.py [hunk 9]
```diff
@@ -876,23 +898,24 @@ def _generate_chunk(
...
-        row with ``pad_id`` (verified empirically on ``GPTNeoXForCausalLM``), so neither caller needs
-        to do post-EOS truncation of the per-chunk return value. ``eos_token_id`` is read from
-        ``self.HF_model.config.eos_token_id`` rather than passed in, since both callers always want the
-        model's configured EOS.
+        stripping convention. The actual engine call is delegated to :attr:`backend` (HF by default;
+        see issue #88). Within a single HF call the engine already fills positions after the first
+        EOS per row with ``pad_id`` (verified empirically on ``GPTNeoXForCausalLM``), so neither
+        caller needs to do post-EOS truncation of the per-chunk return value. ``eos_token_id`` is
+        read from ``self.HF_model.config.eos_token_id`` rather than passed in, since both callers
+        always want the model's configured EOS.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/model.py]: Fixed in 6fa671d. The `Model._generate_chunk` docstring now frames post-EOS padding as a backend-contract requirement (referencing `GenerationBackend`) and names HFBackend as the case that satisfies it for free. Also mirrors the new protocol-level language from the previous comment so the two docstrings stay aligned.

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            input_ids: ``[B, L_in]`` padded prompt tokens. Padding direction is caller-defined;
+                ``attention_mask`` (``True`` at real positions) is what identifies real tokens.
+            attention_mask: Optional ``[B, L_in]`` attention mask; ``True`` for real prompt tokens.
+            generation_config: A fully-populated :class:`transformers.GenerationConfig` describing
+                per-call budget (``max_new_tokens``), sampling mode, pad/EOS ids, etc.
+            **kwargs: Backend-specific per-call options. Callers may provide keys that are only
+                meaningful to some backends (e.g. HF ``logits_processor``); implementations must
+                only forward options supported by the active engine and silently ignore or strip
+                the rest, since e.g. ``transformers.GenerationMixin.generate`` raises on unknown
+                kwargs.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Addressed in a8f5ffc by refining the protocol wording rather than changing `HFBackend`. The tension you identified is real: "silently ignore or strip" was too strict a promise for backends whose engine accepts a broad-and-dynamic kwarg set (HF's `generate` has a VAR_KEYWORD that gets validated at call time).

The updated contract now says implementations "must only forward keys the active engine accepts," and 'accepts' is implementation-defined:
- A backend may strip ahead of time against an explicit allowlist (SGLang will do this in PR 2).
- Or defer to the engine's own runtime validation when the engine's accepted set is broad-and-dynamic (HFBackend's VAR_KEYWORD branch).

In the second case, a truly-unknown key may raise from the engine rather than being silently stripped — explicitly permitted. What's still forbidden is forwarding a key the backend *knows* the engine rejects.

Pushing back on the suggested rewording, though: it drops the cross-backend-dictionary invariant callers rely on. The updated wording keeps the stronger promise: a caller can pass a union dict of kwargs across backends and expect each backend to accept or safely strip its known-unsupported keys. Without that, PR 2's SGLang integration can't pass a shared options dict.

File: src/MEDS_EIC_AR/model/backends/hf.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    """
+
+    def __init__(self, hf_model: PreTrainedModel):
+        self.hf_model = hf_model
+        sig = inspect.signature(hf_model.generate)
+        # If HF ``generate`` has a VAR_KEYWORD (``**kwargs``), don't filter — many valid keys
+        # (``temperature``, ``top_p``, etc.) flow through it rather than being explicit
+        # parameters, and the name-only filter would drop them. HF's own runtime validation
+        # handles truly-unknown kwargs.
+        self._has_var_keyword = any(p.kind is inspect.Parameter.VAR_KEYWORD for p in sig.parameters.values())
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/hf.py:54]: Pushing back: this is a false positive. The line is 109 characters, the project's configured ruff `line-length` is 110 (`pyproject.toml:92`), and pre-commit + the `code-quality` CI check both pass on this line (commit 6fa671d, green). Verified locally:

```
$ awk 'NR==54 {print length($0)}' src/MEDS_EIC_AR/model/backends/hf.py
109
```

No change needed.

File: src/MEDS_EIC_AR/model/backends/base.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+today, SGLang or another engine in the future — see issue #88) from *how* the surrounding rolling
+sliding-window loop, EOS bookkeeping, and token accounting work. Only the innermost
+``generate_chunk`` call is swappable; everything else stays in :class:`MEDS_EIC_AR.model.model.Model`.
+
+A backend implementation takes a padded prompt tensor (this repo pads on the left — see
+``configs/datamodule/generate_trajectories.yaml``'s ``padding_side: LEFT`` — and rolling-chunk
+prompts may additionally contain post-EOS padding on the right for samples that already
+finished) plus the HF-style ``GenerationConfig`` the caller has already built, and returns the
+newly generated tokens (the portion of the HF output that comes *after* the prompt). The
+slicing is deliberately the backend's responsibility: engines like SGLang produce "new tokens
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_EIC_AR/model/backends/base.py]: Good catch — fixed in a8f5ffc. The docstring now points at `src/MEDS_EIC_AR/configs/datamodule/generate_trajectories.yaml`, which is where the config actually lives (verified: no top-level `configs/` directory exists in this repo). Line-wrap on the surrounding paragraph also adjusted to stay under the 110-char ruff limit.
